contracts/autonity/contract/contracts/Autonity.sol

```
@@ -433,6 +433,47 @@ contract Autonity {
 
     }
 
+    /*
+    * getProposer
+    * Implement a stateless weighted-round-robin proposer election from the committee which is in a consistent view
+    * across the nodes, the view is calculated by function computeCommittee() on each block finalization.
+    */
+    function getProposer(uint256 height, uint256 round) public view returns(address) {
+        // if no committee set, return operator address.
+        if (committee.length == 0) {
+            return operatorAccount;
+        }
+        // calculate total voting power from current committee.
+        uint256 total_voting_power = 0;
+        for (uint256 i = 0; i < committee.length; i++) {
+            total_voting_power += committee[i].votingPower;
+        }
+
+        // fallback to round robin if total voting power is 0.
+        if (total_voting_power == 0) {
+            uint256 index = (height + round) % committee.length;
+            return committee[index].addr;
+        }
```

It feels to me like we shouldn't need this because we shouldn't allow validators without stake, is that allowed?

---

> It feels to me like we shouldn't need this because we shouldn't allow validators without stake, is that allowed?

Yes, it is not allow, but don't know if some expection or mistake happens can impact this schdueling, if it happens we cannot on-hold the network but to consider the fallback with round robin.

---

We shouldn't allow invalid states to exist in the system, if it is not allowed and it in fact happens this implies a programming error. Our response to programming errors should be to crash the system immediately since the only solution to a programming error is to change the code.

Having conditions like this delays us finding out about the programming errors, and in the meantime runs the risk that the system is allowed to continue functioning incorrectly, with the ultimate risk in this case that considerable real world value is mishandled or lost.

---

> We shouldn't allow invalid states to exist in the system, if it is not allowed and it in fact happens this implies a programming error. Our response to programming errors should be to crash the system immediately since the only solution to a programming error is to change the code.
> 
> Having conditions like this delays us finding out about the programming errors, and in the meantime runs the risk that the system is allowed to continue functioning incorrectly, with the ultimate risk in this case that considerable real world value is mishandled or lost.

Well, I agree we shouldn't allow invalid states to exist in the system. However we cannot make 100% sure that there isn't an invalid state happens during the run time. So to consider it in pessimism style, why not to check it with those extra condition for a better robustness.

---

>why not to check it with those extra condition for a better robustness.

If we see that we have an invalid state, that means that we have a programming error. The only way to fix this is to stop the system fix the code and fix the state.

Allowing the system to continue running after we know that it contains programming errors is *not* robust. Because once you have reached an invalid state you cannot rely on any of the code in the system to work correctly any more, since when we write code, we assume some valid state for the code to act upon.

What I'm advocating here is failing fast [this](https://www.martinfowler.com/ieeeSoftware/failFast.pdf) article gives a more in depth view on why it is better to fail fast.